File: android/app/src/main/java/com/hexzero/game/plugins/WidgetPlugin.java
package com.hexzero.game.plugins;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.hexzero.game.HexZeroWidgetProvider;

@CapacitorPlugin(name = "Widget")
public class WidgetPlugin extends Plugin {

    @PluginMethod
    public void updateWidgetData(PluginCall call) {
        Integer gamesPlayed = call.getInt("gamesPlayed");
        Integer gamesWon = call.getInt("gamesWon");
        Integer currentStreak = call.getInt("currentStreak");

        if (gamesPlayed == null || gamesWon == null || currentStreak == null) {
            call.reject("Missing required parameters");
            return;
        }

        // Update widget data through the widget provider
        HexZeroWidgetProvider.updateWidgetData(
            getContext(),
            gamesPlayed,
            gamesWon,
            currentStreak
        );

        call.resolve();
    }
}

File: ios/App/App/Plugins/WidgetPlugin.swift
import Capacitor
import WidgetKit

@objc(WidgetPlugin)
public class WidgetPlugin: CAPPlugin {

    @objc func updateWidgetData(_ call: CAPPluginCall) {
        guard let gamesPlayed = call.getInt("gamesPlayed"),
              let gamesWon = call.getInt("gamesWon"),
              let currentStreak = call.getInt("currentStreak") else {
            call.reject("Missing required parameters")
            return
        }

        let lastPlayed = call.getDouble("lastPlayed") ?? Date().timeIntervalSince1970

        // Save to shared UserDefaults (app group)
        if let sharedDefaults = UserDefaults(suiteName: "group.com.hexzero.game") {
            sharedDefaults.set(gamesPlayed, forKey: "widget_gamesPlayed")
            sharedDefaults.set(gamesWon, forKey: "widget_gamesWon")
            sharedDefaults.set(currentStreak, forKey: "widget_currentStreak")
            sharedDefaults.set(lastPlayed, forKey: "widget_lastPlayed")
            sharedDefaults.synchronize()

            // Request widget reload on iOS 14+
            if #available(iOS 14.0, *) {
                WidgetCenter.shared.reloadAllTimelines()
            }

            call.resolve()
        } else {
            call.reject("Failed to access shared UserDefaults")
        }
    }
}
